src/rt/sched/backpressure.h

```diff
@@ -0,0 +1,249 @@
+ * to it will be temporarily descheduled ("muted") so that the overloaded cown
+ * can work through its queue. Similarly, cowns sending to muted receivers will
+ * also be muted so that cowns do not themselves become overloaded as a
+ * consequence of being muted. Muted cowns will be rescheduled after the
+ * receiver that triggered their muting (the "mutor") is no longer muted or
+ * overloaded.
+ *
+ * ## Detecting Overloaded Cowns
+ *
+ * An overloaded cown is one that experiences message queue growth as it is
+ * processing messages. Since it would be prohibitively expensive to calculate
+ * the length of a message queue directly, a count of messages removed from the
+ * queue over some period is used as an approximation for message queue size.
+ * Like scheduler fairness, the period is determined using a token message that
+ * will occasionally fall out of the queue and then be inserted to begin the
+ * next period. The "load" of a cown at any time is the accumulation of messages
+ * processed over multiple periods.
+ *
+ * ## Muting Cowns
+ *
+ * All messages from one set of sending cowns to a set of receiving cowns are
+ * scanned to determine if the senders should be muted. The senders will be
+ * muted once their action completes if all of the following are true:
+ *   1. The receiver set does not contain any of the senders.
+ *   2. No sender is overloaded.
+ *   3. Any of the receivers are either overloaded or muted.
+ *
+ * A message is given "priority" if either condition 1 or 2 are false. A
+ * priority message will result in any muted receivers being unmuted so that the
+ * senders may make progress sooner.
+ *
+ * If all of the above conditions are met, the first receiver that is either
+ * overloaded or muted is identified as the mutor for the senders. First in this
+ * case means the least cown, determined by the mutlimessage cown order.
+ *
+ * ## Tracking Muted Cowns
+ *
+ * Once a scheduler thread completes a message action that would result in
+ * muting the cowns running the messsage, the cowns are acquired by the
+ * scheduler thread and placed in a "mute map" on the scheduler thread. The mute
+ * map is a mapping from mutor => mute set, where "mute set" refers to the set
+ * of cowns muted by a mutor.
+ *
+ * A mutor may exist as a key in mutiple mute maps, since cowns may continue
+ * sending messages to it on other threads and then be subsequently muted by
+ * that thread on which the senders were running. However, a muted cown must
+ * only exist in a mute set on the scheduler thread that muted it.
+ *
+ * A muted cown may not be scheduled or collected until they are removed from
+ * their scheduler thread's mute map and rescheduled by that thread ("unmuted").
+ *
+ * Each scheduler thread scans the mutors of its mute map on each iteration of
+ * the run loop. If any mutor is not muted and is not overloaded, all cowns in
+ * the corresponding mute set are unmuted and the entry is removed from the mute
+ * map.
+ */
```

Comment:
I think you should add either here or near the beginning, some examples of cases where backpressure will work to control unbounded queue growth and example cases where it will not.

Reply:
I've added a section for known limitations